src/compile/legend/parse.ts

```diff
@@ -49,9 +53,12 @@ function getLegendDefWithScale(model: UnitModel, channel: Channel): VgLegend {
 }
 
 export function parseLegendForChannel(model: UnitModel, channel: NonspatialScaleChannel): LegendComponent {
-  const fieldDef = model.fieldDef(channel);
   const legend = model.legend(channel);
+  const fieldDef = model.fieldDef(channel);
 
+  if (!model.getScaleComponent(channel)) {
```

Comment:
It does. 

I mean move the logic to `parseUnitLegend` so that the structure of both `parseUnitLegend` and `parseUnitAxis` are similar.